### ISSUE TITLE
storage: replicas not balanced across multiple stores in same node

### ISSUE THREAD

**Issue by cuongdo** [cockroachdb/cockroach#6782]
When running a cluster where nodes have multiple stores, one store often seems to get a disproportionate number of replicas allocated to it.

Here's an example cluster with 5 nodes, each with 5 stores:

<img width="1219" alt="screenshot 2016-05-18 16 51 54" src="https://cloud.githubusercontent.com/assets/1564582/15374797/0105ad8c-1d1a-11e6-8192-fe3f31e2e9b9.png">

This is happening on 3 of 5 nodes in this cluster. We currently do not rebalance between multiple stores in the same node, but we may want to consider it for performance reasons. For example, maybe each store is backed by a different SSD. Spreading our I/O evenly across stores could benefit performance.

**Comment by bdarnell** [comment]
Moving replicas between stores on the same node is tracked in #2067 (and i'll add some more comments there), but I think this is a separate issue. Those ranges should be moved *somewhere*, even if it's to an underloaded store on another node.

**Comment by tschottdorf** [comment]
Closing as stale.

**Comment by a-robinson** [comment]
I get that we're very unlikely to work on this anytime soon, but it's a legitimate flaw in the product, not just a feature request. Do we not keep that sort of thing around anymore?

**Comment by tschottdorf** [comment]
Oh, I missed that facet. I thought this was just a long-fixed replication bug. Sorry and thanks for complaining. Lots of issues to look into.

**Comment by knz** [comment]
Update from @tbg: with the recent changes to allow atomic replication changes, we're moving closer to resolving this issue. However:

> This doesn’t work yet, but only because it’s out of scope. The work here is to remove the check that prevents us from doing it, and making sure there aren’t any other places where we’ve since bought into the “no two replicas of a range on a node” pattern
> I think for the second part an audit on anyone who accesses NodeID should be enough. We want to make sure nobody tries to resolve a replica via its NodeID

Also a casual search also reveals related issue #39415 which we need to solve (maybe via PR #39497 from @nvanbenschoten) which will likely need to be solved before we can consider multi-store nodes safe for use.


(cc @johnrk as this is an important long-standing issue which we're aiming to solve soon-ish)

**Comment by nvanbenschoten** [comment]
@lunevalex this is the issue I was thinking about. This has been unblocked by https://github.com/cockroachdb/cockroach/issues/12768.

Most of the work here will be in the `kvserver.Allocator`, where we need to lift the restriction about rebalancing between stores on the same node if we can rebalance atomically between them. I haven't personally touched this code in a long time, so @tbg might be better able to advise you on the details and subtleties that will go into addressing this.

**Comment by tbg** [comment]
I would start here by trying out a smoke test that proves that #12768 really fundamentally does allow the lateral rebalance. Concretely, I think this would be a test with boilerplate similar to

https://github.com/cockroachdb/cockroach/blob/f80ec9c292d43676d739fffabc847fb8b3195396/pkg/kv/kvserver/client_atomic_membership_change_test.go#L34

except that we want to start only a single server, with two stores:

https://github.com/cockroachdb/cockroach/blob/7b7fcb0830b164b874ee36f238d2667804c33550/pkg/base/test_server_args.go#L66

and then

```go
	desc = runChange(desc, []roachpb.ReplicationChange{
		{ChangeType: roachpb.ADD_REPLICA, Target: roachpb.ReplicationTarge{NodeID:1 StoreID: 2},
		{ChangeType: roachpb.REMOVE_REPLICA, Target: roachpb.ReplicationTarge{NodeID:1 StoreID: 1},
	})
```

This should work (after removing some checks that probably refuse this for legacy reasons).

Assuming this is in the bag, we can turn our attention to the replicate queue. Rebalances generally start out here:

https://github.com/cockroachdb/cockroach/blob/76c63ded01f18aa31c08ed6a2e9cf4972b0b0459/pkg/kv/kvserver/replicate_queue.go#L412

We should ultimately be able to end up here with replication targets similar to the test:

https://github.com/cockroachdb/cockroach/blob/f291061a757203d852500bb9cbc2fefa59b68deb/pkg/kv/kvserver/allocator.go#L748-L755

I browsed through the code and the only code in the allocator that rules out lateral rebalances now is here:

https://github.com/cockroachdb/cockroach/blob/4cf5a971449c843b8aff9e5cdab6f4a315bc35d5/pkg/kv/kvserver/allocator_scorer.go#L613-L621

We can't just remove that - we want to allow a second replica on the same node only if we are in the context of a lateral move (we would not want to add a second replica to the node without atomically removing the first one). This will need to be established and unit tested.

Finally, we should have some sort of high-level validation that this works end-to-end. For example, we could start a three-node cluster on which the first node has two stores and ensure that we organically see replicas move to that store (which means there must have been lateral rebalances).

I'll point out that there are likely deficiencies in how the system tries to balance the load. For example, if n1 has 100 stores and n2 and n3 have one, then effectively n1 will end up with 100/102 percent of data (I think) which may not be what we want (n1 might be CPU bound long before that). But for mostly symmetric configurations, hopefully it would do the right thing. While we're looking at that code, we should document as much as possible the expected behavior and its strengths. I believe we have some simulated rebalancing unit tests, we should update those.

**Comment by johnrk** [comment]
We have access to a VMware test environment until September. We should consider using that environment to test with [vSAN](https://docs.vmware.com/en/VMware-vSAN/index.html).
